### PR TITLE
Feature/character switching

### LINKED ISSUE #63: Switching Character via keybinding

### PR DESCRIPTION
**Pull request by ColinT**
Closes #63 

This is currently not ready for merge. I am just putting this on the radar since the `dev` branch is active to prevent overlapping development.

This PR introduces support for switching characters via keyboard and XInput gamepads.

Global keyboard shortcuts are supported via Electron's [`globalShortcut`](https://www.electronjs.org/docs/api/global-shortcut).
Local keyboard shortcuts are supported via [`electron-localshortcut`](https://www.npmjs.com/package/electron-localshortcut).
Gamepad supported is provided via the [Gamepad API](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad_API).

Also, is there an auto-formatter being used for this project? My personal code style is significantly different from the source. If you provide me a guideline or settings I will happily configure my formatter to comply.

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -12,134 +12,141 @@
 "@babel/code-frame@^7.0.0-beta.35":
   version "7.0.0"
   resolved "https://registry.yarnpkg.com/@babel/code-frame/-/code-frame-7.0.0.tgz#06e2ab19bdb535385559aabb5ba59729482800f8"
-  integrity sha512-OfC2uemaknXr87bdLUkWog7nYuliM9Ij5HUcajsVcMCpQrcLmtxRbVFTIqmcSkSeYRBFBRxs2FiUqFJDLdiebA==
```
Comment (Tarnadas) [inline yarn.lock]: you probably need to update yarn

**Comment by ColinT** [comment]
I've finished writing the features I wanted implemented. This now includes future support for combination and multiple hotkeys for the same shortcut. All non-modifier Electron accelerators are now supported, instead of just alphanumeric keys.

I have fixed all relevant `eslint` errors with one major exception:
`42:17  error  Do not use findDOMNode  react/no-find-dom-node`
https://github.com/ColinT/net64plus/blob/229e3feb4d8f9c99a17da87b99d1596694060b36/src/renderer/components/buttons/HotkeyButton.tsx#L42

`HotkeyButton` relies on `findDOMNode` to detect when a click is performed outside of its node, and deactivate. This was the best solution I could find, but I am not that familiar with React. Feel free to make a suggestion for this if you have one.

This PR is fairly hefty so feel free to ask for any clarifications or additional inline comments should you feel it to be beneficial.

File: proto/ClientServerMessage.js [hunk 2]
```diff
@@ -2054,7 +2054,13 @@ $root.PlayerBytes = (function() {
         var object = {};
         if (options.defaults) {
             object.playerId = 0;
-            object.playerData = options.bytes === String ? "" : [];
```
[the remark this answers is not in the document]
Comment (ColinT) [reply proto/ClientServerMessage.js]: These were generated when I built the project.

I'm not familiar with your build pipe, but I can remove the changes to the `/proto` files from the PR.

File: src/renderer/components/buttons/HotkeyButton.tsx [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+    this.keyDownListener = this.keyDownListener.bind(this)
+    this.buttonDownListener = this.buttonDownListener.bind(this)
+    this.clickOutsideListener = this.clickOutsideListener.bind(this)
+    this.removeListeners = this.removeListeners.bind(this)
+  }
+
+  private node: Node | undefined
+
+  componentDidMount () {
+    this.node = ReactDOM.findDOMNode(this) as Node
```
[the remark this answers is not in the document]
Comment (ColinT) [reply src/renderer/components/buttons/HotkeyButton.tsx]: Thanks for the reference. I found a solution using `React.createRef()`.

File: src/main/HotkeyManager.ts [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+export class HotkeyManager {
+  private _username = ''
+  public set username(value: string) { this._username = value }
+  public get username(): string { return this._username }
+
+  private _characterCyclingOrder: Array<{characterId: number, on: boolean}> = []
+  private _characterCyclingIndex: number = 0;
+
+  private _hotkeyBindings: { [shortcut: string]: string | undefined } = {}
```
[the remark this answers is not in the document]
Comment (ColinT) [reply src/main/HotkeyManager.ts]: Thanks. I made this change, but `Object.entries` and `Object.keys` still transforms the enum keys into a string.

Not sure I can iterate over an enum type.

**Pull request merged**